lib/AST/CanonBounds.cpp

```diff
@@ -0,0 +1,532 @@
+     // case: ExtVectorElementExpr
+     case Expr::VAArgExprClass: break;
+     case Expr::GenericSelectionExprClass: Cmp = CompareGenericSelectionExpr(E1, E2); break;
+
+     // Atomic expressions.
+     case Expr::AtomicExprClass: CompareAtomicExpr(E1, E2); break;
+
+     // GNU Extensions.
+     case Expr::AddrLabelExprClass: break;
+     case Expr::StmtExprClass: break;
+     case Expr::ChooseExprClass: break;
+     case Expr::GNUNullExprClass: break;
+
+     // Bounds expressions
+     case Expr::NullaryBoundsExprClass: Cmp = CompareNullaryBoundsExpr(E1, E2); break;
+     case Expr::CountBoundsExprClass: Cmp = CompareCountBoundsExpr(E1, E2); break;
+     case Expr::RangeBoundsExprClass:  Cmp = CompareRangeBoundsExpr(E1, E2); break; break;
+     case Expr::InteropTypeBoundsAnnotationClass: Cmp = CompareInteropTypeBoundsAnnotation(E1, E2); break;
+     case Expr::PositionalParameterExprClass: Cmp = ComparePositionalParameterExpr(E1, E2); break;
+     case Expr::BoundsCastExprClass: Cmp = CompareBoundsCastExpr(E1, E2); break;
+
+     // Clang extensions
+     case Expr::ShuffleVectorExprClass: break;
+     case Expr::ConvertVectorExprClass: break;
+     case Expr::BlockExprClass: Cmp = CompareBlockExpr(E1, E2); break;
+     case Expr::OpaqueValueExprClass: break;
+     case Expr::TypoExprClass: break;
+     // TODO:
+     // case Expr::MSPropertyRefExprClass:
+     // case Expr::MSPropertySubscriptExprClass:
+
+     default:
+       llvm_unreachable("unexpected expression kind");         
+   }
+
+   if (Cmp != Result::Equal)
+     return Cmp;
+
+   // Check children
+   auto I1 = E1->child_begin();
+   auto I2 = E2->child_begin();
+   for ( ;  I1 != E1->child_end() && I2 != E2->child_end(); ++I1, ++I2) {
+     const Stmt *E1Child = *I1;
+     const Stmt *E2Child = *I2;
+     bool ordered = false;
+     Cmp = ComparePointers(E1Child, E2Child, ordered);
+     if (ordered) {
+       if (Cmp == Result::Equal)
+         continue;
+       return Cmp;
+      }
+     assert(E1Child && E2Child);
+     const Expr *E1ChildExpr = dyn_cast<Expr>(E1Child);
+     const Expr *E2ChildExpr = dyn_cast<Expr>(E2Child);
+     assert(E1ChildExpr && E2ChildExpr && "dyn_cast failed");
+     if (E1ChildExpr && E2ChildExpr) {
+       Cmp = CompareExpr(E1ChildExpr, E2ChildExpr);
+       if (Cmp != Result::Equal)
+         return Cmp;
+     } else
+       // assert fired - return something
+       return Result::LessThan;
+   }
+
+   // Make sure the number of children was equal.  If E1 has
+   // fewer children than E2, make it less than E2.  If it has more
+   // children, make it greater than E2.
+   if (I1 == E1->child_end() && I2 != E2->child_end())
+     return Result::LessThan;
+
+   if (I1 != E1->child_end() && I2 == E2->child_end())
+     return Result::GreaterThan;
+
+   return Lexicographic::Result::Equal;
+}
+
+// Lexicographic comparion of properties specific to each expression type.
+// Does not check children of expressions.
+#define CHECK_WRAPPER(N) \
```

Comment:
This is definitely going to catch someone out, that this opens but does not close a scope. I don't know how to do it better though. Maybe take 2 arguments, the second one being a block, like so:

```C
#define CHECK_WRAPPER(N, Block) \
Lexicographic::Result \
 Lexicographic::Compare##N(const Expr *Raw1, const Expr *Raw2) { \
   const N *E1 = dyn_cast<N>(Raw1); \
   const N *E2 = dyn_cast<N>(Raw2); \
   if (!E1 || !E2)  { \
     llvm_unreachable("dyn_cast failed"); \
     return Result::LessThan; \
   }
   Block
}

// Then used like so:
CHECK_WRAPPER(DeclRefExpr, return CompareDecl(E1->getDecl(), E2->getDecl());)

// Or:
CHECK_WRAPPER(IntegerLiteral, {
  BuiltinType::Kind Kind1 = E1->getType()->castAs<BuiltinType>()->getKind();
  BuiltinType::Kind Kind2 = E2->getType()->castAs<BuiltinType>()->getKind();
  Result Cmp = CompareInteger(Kind1, Kind2);
  if (Cmp != Result::Equal)
    return Cmp;
  return CompareAPInt(E1->getValue(), E2->getValue());
})
```